lib/datadog/statsd.rb

```diff
@@ -103,36 +87,31 @@ def initialize(
         raise ArgumentError, 'tags must be a Array<String> or a Hash'
       end
 
-      tags = tag_hash_to_array(tags) if tags.is_a?(Hash)
-      @tags = (tags || []).compact.map! do |tag|
-        escape_tag_content(tag)
-      end
+      @namespace = namespace
+      @prefix = @namespace ? "#{@namespace}.".freeze : nil
 
-      # append the entity id to tags if DD_ENTITY_ID env var is not nil
-      unless ENV.fetch('DD_ENTITY_ID', nil).nil?
-        dd_entity = escape_tag_content(ENV.fetch('DD_ENTITY_ID', nil))
-        @tags << 'dd.internal.entity_id:' + dd_entity
-      end
+      @serializer = Serialization::Serializer.new(prefix: @prefix, global_tags: tags)
 
-      # init telemetry
-      transport_type = socket_path.nil? ? 'udp': 'uds'
-      telemetry_tags = (["client:ruby", "client_version:#{VERSION}", "client_transport:#{transport_type}"] + @tags).join(COMMA).freeze
-      @telemetry = Telemetry.new(disable_telemetry, telemetry_tags, telemetry_flush_interval)
+      transport_type = socket_path.nil? ? :udp : :uds
 
-      if socket_path.nil?
-        @connection = UDPConnection.new(host, port, logger, @telemetry)
-      else
-        @connection = UDSConnection.new(socket_path, logger, @telemetry)
-      end
-      @logger = logger
+      @telemetry = Telemetry.new(disable_telemetry, telemetry_flush_interval,
+        global_tags: tags,
+        transport_type: transport_type
+      )
 
-      @namespace = namespace
-      @prefix = @namespace ? "#{@namespace}.".freeze : nil
+      @connection = case transport_type
+                    when :udp
+                      UDPConnection.new(host, port, logger, telemetry)
+                    when :uds
+                      UDSConnection.new(socket_path, logger, telemetry)
+                    end
+
+      @logger = logger
 
       @sample_rate = sample_rate
 
       # we reduce max_buffer_bytes by a the rough estimate of the telemetry payload
-      @batch = Batch.new(@connection, (max_buffer_bytes - @telemetry.estimate_max_size))
+      @batch = Batch.new(@connection, (max_buffer_bytes - telemetry.estimate_max_size))
```

Comment:
Why is using the `telemetry` accessor better than `@telemetry` like before?
And, in that case, why not using also the `connection` accessor?

Reply:
I like to use getters because the error are clearer in case of a typo.
e.g.
```
irb(main):006:0> an_aattribute.to_s
Traceback (most recent call last):
        2: from /Users/karim.bogtob/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        1: from (irb):6
NameError (undefined local variable or method `an_aattribute' for main:Object)
Did you mean?  an_attribute
```

Wheras with a @ we would get:
```
irb(main):008:0> @an_aattribute.to_s
=> ""
```
Because an unknown method raise an error whereas n unknown attribute is replaced by nil.